File: Stacks & Queues/Stack.js
// STACKS WITH LINKEDLIST

class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
}
class LinkedStack {
    constructor() {
        this.top = null;
        this.bottom = null;
        this.length = 0;
    }

    peek() {
        return this.top;

    }
    push(value) {
       const newNode  = new Node(value)
       newNode.value = value 
       if (this.length === 0) {
            this.top = newNode;
            this.bottom = newNode;
        } else {
            const holdingPotinter = this.top;
            this.top = newNode
            this.top.next = holdingPotinter;
        }
        this.length++;
        return this;
    }
    pop() {
        if (!this.top) {
            return null;
        }
        if (this.length === 1) {
            this.top = null;
            this.bottom = null;
        }
        this.top = this.top.next;
        this.length--;
        return this

    }
}



// TESTING

const myStack = new LinkedStack();
console.log(myStack.peek());
    myStack.push("google");
    myStack.push("udemy");
    myStack.push("discord");
    console.log(myStack.peek());
    console.log(myStack.pop());
    console.log(myStack.pop());
myStack.pop();
myStack.pop();


// STACKS WITH ARRAYS


class ArrayStack {
    constructor() {
        this.array = [];
    }

    peek(){
        return(this.array[this.array.length-1])
    }

    push(value) {
        this.array.push(value);
        return this;
    }

    pop() {
        this.array.pop();
        return this;
    }
}
 
// TESTING 

const newStack = new ArrayStack();
console.log(newStack.peek());
console.log(newStack.push("google"));
console.log(newStack.push("udemy"));
console.log(newStack.push("discord"));
console.log(newStack.peek());
console.log(newStack.pop());
console.log(newStack.pop());
console.log(newStack.peek()); 

File: Stacks & Queues/Stack.py
# STACKS WITH LINKED-LIST

class Node:
    def __init__ (self, value):
        self.value = value
        self.next = None



class LinkedStacks:
    def __init__ (self):
        self.top = None
        self.length = 0

    def isEmpty(self):
        if self.length == 0:
            return True # Meaning stack is empty
        else:
            return False # Stack is not empty

    def peek(self):
        if self.isEmpty() == True:
            return "Stack is Empty"
        else:
            return self.top

    def push(self, value):
        
        if self.top is None:
            self.top = Node(value)
        else: 
            newNode = Node(value)
            # newNode.next = self.top
            self.top.next = newNode
            self.top = newNode
        
        self.length +=1
        print(self.top.value)
        return self.top.value

    def pop(self):
        if not self.top:
            return (None)
        elif self.length == 1:
            self.top = None
        else:
            popped_node = self.top
            self.top = self.top.next
            popped_node.next = None
        self.length -=1
        return self.top.value
    
    
 




# TESTING

# myStack = LinkedStacks()
# print(myStack.peek())
# print(myStack.push("google"))
# print(myStack.push("udemy"))
# print(myStack.push("discord"))
# print(myStack.pop())
# print(myStack.pop())
# print(myStack.peek())

# STACKS WITH ARRAYS

class ArraysStack:
    def __init__(self):
        self.array = []

    def peek(self):
        if len(self.array) == 0:
            return ("Stack is Empty")
        else:
            return self.array[-1]


    def push(self, value):
        self.array.append(value)
        return self.array
    
    def pop(self):
        self.array.pop()
        return self.array

# TESTING

newStack = ArraysStack()
print(newStack.peek())
print(newStack.push("google"))
print(newStack.push("facebook"))
print(newStack.push("tiktok"))
print(newStack.push("microsoft"))
print(newStack.pop())
print(newStack.pop())
print(newStack.pop())
print(newStack.peek())

